File: Source/Game/WolfArea.cs
using System.Collections.Generic;
using WCSharp.Api;
using WCSharp.Shared.Data;

public class WolfArea
{
    public static Dictionary<int, WolfArea> WolfAreas { get; } = new();
    public static float TotalArea { get; private set; } = 0.0f;
    public int ID { get; set; }
    public rect Rect { get; set; }
    public region Region { get; set; }
    public Rectangle Rectangle { get; set; }
    public float Area { get; private set; }
    public bool IsEnabled { get; set; } = true;
    private trigger AreaTrigger { get; set; }
    public List<Wolf> Wolves { get; set; } = new List<Wolf>();
    public int FixationCount { get; set; }

    public WolfArea(int id, region region)
    {
        ID = id;
        Region = region;
    }

    public static void Initialize()
    {
        int count = 0;
        foreach (var wolfRegion in RegionList.WolfRegions)
        {
            var wolfArea = new WolfArea(count, wolfRegion.Region)
            {
                Rect = wolfRegion.Rect,
                Rectangle = wolfRegion
            };
            wolfArea.CalculateArea();
            wolfArea.RegisterEnterEvents();
            wolfArea.RegisterLeaveEvents();
            WolfAreas.Add(count, wolfArea);
            count++;
        }
    }

    private void RegisterEnterEvents()
    {
        AreaTrigger = trigger.Create();
        AreaTrigger.RegisterEnterRegion(Region, FilterList.KittyFilter);
        AreaTrigger.AddAction(() =>
        {
            try
            {
                var unit = @event.Unit;
                var player = unit.Owner;

                var kitty = Globals.ALL_KITTIES[player];
                kitty.ProgressHelper.CurrentPoint = ID;
                kitty.ProgressZone = ID;
            }
            catch (System.Exception e)
            {
                Logger.Warning($"Error in WolfArea.RegisterEnterEvents: {e.Message}");
                throw;
            }
        });
    }

    /// <summary>
    /// Prevents wolves from leaving the area with wander.
    /// </summary>
    private void RegisterLeaveEvents()
    {
        AreaTrigger = trigger.Create();
        AreaTrigger.RegisterLeaveRegion(Region, FilterList.DogFilter);
        AreaTrigger.AddAction(() =>
        {
            try
            {
                var wolf = Globals.ALL_WOLVES[@event.Unit];
                wolf.WolfMove();
            }
            catch (System.Exception e)
            {
                Logger.Critical($"Error in WolfArea.RegisterLeaveEvents: {e.Message}");
            }
        });
    }

    private void CalculateArea()
    {
        Area = Rectangle.Width * Rectangle.Height;
        TotalArea += Area;
    }
}

File: src/Game/WolfArea.ts
import { Logger } from 'src/Events/Logger/Logger'
import { Globals } from 'src/Global/Globals'
import { RegionList } from 'src/Global/RegionList'
import { FilterList } from 'src/Utility/FilterList'
import { getTriggerUnit } from 'src/Utility/w3tsUtils'
import { Rectangle, Trigger } from 'w3ts'
import { Wolf } from './Entities/Wolf'

export class WolfArea {
    public static WolfAreas: Map<number, WolfArea> = new Map<number, WolfArea>()
    public static TotalArea = 0.0
    public ID = 0
    public Rect: rect
    public Region: region
    public Rectangle: Rectangle
    public Area = 0
    public IsEnabled: boolean = true
    private AreaTrigger: Trigger
    public Wolves: Wolf[] = []
    public FixationCount = 0

    public constructor(id: number, region: region) {
        this.ID = id
        this.Region = region
    }

    public static Initialize = () => {
        let count = 0
        for (const wolfRegion of RegionList.WolfRegions) {
            const wolfArea = new WolfArea(count, wolfRegion.region())
            wolfArea.Rect = Rect(wolfRegion.minX, wolfRegion.minY, wolfRegion.maxX, wolfRegion.maxY)
            wolfArea.Rectangle = wolfRegion
            wolfArea.CalculateArea()
            wolfArea.RegisterEnterEvents()
            wolfArea.RegisterLeaveEvents()
            WolfArea.WolfAreas.set(count, wolfArea)
            count++
        }
    }

    private RegisterEnterEvents = () => {
        const AreaTrigger = Trigger.create()!
        AreaTrigger.registerEnterRegion(this.Region, FilterList.KittyFilter)
        AreaTrigger.addAction(() => {
            try {
                const unit = getTriggerUnit()
                const player = unit.owner

                const kitty = Globals.ALL_KITTIES.get(player)!
                kitty.ProgressHelper.CurrentPoint = this.ID
                kitty.ProgressZone = this.ID
            } catch (e: any) {
                Logger.Warning(`Error in WolfArea.RegisterEnterEvents: ${e}`)
                throw e
            }
        })
    }

    /// <summary>
    /// Prevents wolves from leaving the area with wander.
    /// </summary>
    private RegisterLeaveEvents = () => {
        const AreaTrigger = Trigger.create()!
        AreaTrigger.registerLeaveRegion(this.Region, FilterList.DogFilter)
        AreaTrigger.addAction(() => {
            try {
                const wolf = Globals.ALL_WOLVES.get(getTriggerUnit())!
                wolf.WolfMove()
            } catch (e: any) {
                Logger.Critical(`Error in WolfArea.RegisterLeaveEvents: ${e}`)
            }
        })
    }

    private CalculateArea = () => {
        const width = Math.abs(this.Rectangle.maxX - this.Rectangle.minX)
        const height = Math.abs(this.Rectangle.maxY - this.Rectangle.minY)

        this.Area = width * height
        WolfArea.TotalArea += this.Area
    }
}
